porch/controllers/workloadidentitybindings/pkg/controllers/workloadidentitybinding/controller.go

```diff
@@ -213,9 +245,115 @@ func (r *WorkloadIdentityBindingReconciler) BuildObjectsToApply(ctx context.Cont
 		})
 		obj.SetOwnerReferences(ownerRefs)
 	}
+
 	return objects, nil
 }
 
+func (r *WorkloadIdentityBindingReconciler) addWiAnnotation(ctx context.Context, projectID string, subject *api.WorkloadIdentityBinding) error {
+	// TODO: We should have a watch here so we can annotate the ServiceAccount even if it is
+	// applied later.
+	ksa, err := r.getKubernetesServiceAccount(ctx, subject)
+	if err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	// TODO: Same as above, we should have a watch here so we can annotate the ksa when
+	// the gsa is created and reconciled (i.e. have the status.email field set)
+	gsa, err := r.getGcpServiceAccount(ctx, subject)
+	if err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	gsaEmail, found, err := unstructured.NestedString(gsa.Object, "status", "email")
+	if err != nil {
+		return err
+	}
+	if !found || gsaEmail == "" {
+		return nil
+	}
+
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("ServiceAccount"))
+	u.SetName(ksa.GetName())
+	u.SetNamespace(ksa.GetNamespace())
+	u.SetAnnotations(map[string]string{
+		"iam.gke.io/gcp-service-account": gsaEmail,
+	})
+
+	return r.updateServiceAccount(ctx, u, subject)
+}
+
+func (r *WorkloadIdentityBindingReconciler) removeWiAnnotation(ctx context.Context, subject *api.WorkloadIdentityBinding) error {
+	sa, err := r.getKubernetesServiceAccount(ctx, subject)
+	if err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("ServiceAccount"))
+	u.SetName(sa.GetName())
+	u.SetNamespace(sa.GetNamespace())
+	u.SetAnnotations(make(map[string]string))
+
+	return r.updateServiceAccount(ctx, u, subject)
+}
+
+func (r *WorkloadIdentityBindingReconciler) getKubernetesServiceAccount(ctx context.Context, subject *api.WorkloadIdentityBinding) (*unstructured.Unstructured, error) {
+	saRef := subject.Spec.KubernetesServiceAccountRef
+
+	saName := saRef.Name
+	saNamespace := saRef.Namespace
+	if saNamespace == "" {
+		saNamespace = subject.GetNamespace()
+	}
+
+	var sa unstructured.Unstructured
+	sa.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("ServiceAccount"))
```

Comment:
Oooh this is a nice way of setting apiVersion/kind that I am going to steal shamelessly :-)